x/stakeibc/keeper/msg_server_create_trade_route.go

```diff
@@ -2,81 +2,187 @@ package keeper
 
 import (
 	"context"
-	"fmt"
+
+	sdkmath "cosmossdk.io/math"
 
 	"github.com/Stride-Labs/stride/v16/x/stakeibc/types"
 
 	errorsmod "cosmossdk.io/errors"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	icatypes "github.com/cosmos/ibc-go/v7/modules/apps/27-interchain-accounts/types"
+	connectiontypes "github.com/cosmos/ibc-go/v7/modules/core/03-connection/types"
 )
 
-func (k msgServer) CreateTradeRoute(goCtx context.Context, msg *types.MsgCreateTradeRoute) (*types.MsgCreateTradeRouteResponse, error) {
+// Gov tx to to register a trade route that swaps reward tokens for a different denom
+//
+// Example proposal:
+//
+//		{
+//		   "title": "Create a new trade route for host chain X",
+//		   "metadata": "Create a new trade route for host chain X",
+//		   "summary": "Create a new trade route for host chain X",
+//		   "messages":[
+//		      {
+//		         "@type": "/stride.stakeibc.MsgCreateTradeRoute",
+//		         "authority": "stride10d07y265gmmuvt4z0w9aw880jnsr700jefnezl",
+//
+//		         "stride_to_host_connection_id": "connection-0",
+//				 "stride_to_reward_connection_id": "connection-1",
+//			     "stride_to_trade_connection_id": "connection-2",
+//
+//				 "host_to_reward_transfer_channel_id": "channel-0",
+//				 "reward_to_trade_transfer_channel_id": "channel-1",
+//			     "trade_to_host_transfer_channel_id": "channel-2",
+//
+//				 "reward_denom_on_host": "ibc/rewardTokenXXX",
+//				 "reward_denom_on_reward": "rewardToken",
+//				 "reward_denom_on_trade": "ibc/rewardTokenYYY",
+//				 "host_denom_on_trade": "ibc/hostTokenZZZ",
+//				 "host_denom_on_host": "hostToken",
+//
+//				 "pool_id": 1,
+//				 "min_swap_amount": 10000000,
+//				 "max_swap_amount": 1000000000
+//			  }
+//		   ],
+//		   "deposit": "2000000000ustrd"
+//	   }
+//
+// >>> strided tx gov submit-proposal {proposal_file.json} --from wallet
+func (ms msgServer) CreateTradeRoute(goCtx context.Context, msg *types.MsgCreateTradeRoute) (*types.MsgCreateTradeRouteResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
-	k.Logger(ctx).Info(fmt.Sprintf("create trade route: %s", msg.String()))
+	if ms.authority != msg.Authority {
+		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", ms.authority, msg.Authority)
+	}
+
+	// Validate trade route does not already exist for this denom
+	_, found := ms.Keeper.GetTradeRoute(ctx, msg.RewardDenomOnHost, msg.HostDenomOnHost)
```

Comment:
Reasonable check for now, in the future if we allow more than one DEX we could allow multiple traderoutes for the same start and end denom, we would compare the best recent price to decide which one we want.  In that case this check would become more complex for also including the trade chain id... I know this problem isn't only here in this check, it would require changing the keepers as well.

Reply:
yeah good point, but I'd imagine we'll move to the auction approach before we're in a scenario where we want to support multiple DEXs (low confidence though)

Either way, there'd be a bit of a refactor if we did decide to support multiple